File: Stacks/11/remMinValid.js
var minRemoveToMakeValid = function (s) {
  let i = 0;
  const l = s.length;
  if (l === 0) return s;
  const stack = [];
  const invalid = new Set();
  let valid = '';

  while (i < l) {
    let ch = s[i];

    if (ch === '(') {
      stack.push(i);
    } else if (ch === ')') {
      if (stack.length > 0) stack.pop();
      else invalid.add(i);
    }
    i++;
  }

  while (stack.length > 0) {
    invalid.add(stack.pop());
  }

  i = 0;
  while (i < l) {
    if (!invalid.has(i)) valid += s[i];
    i++;
  }

  return valid;
};

File: Stacks/11/remMinValid.py
def minRemoveToMakeValid(s):
    i = 0
    l = len(s)
    if l == 0: return s
    valid = ""
    invalid = set()
    stack = []
    while i < l:
        ch = s[i]
        if ch == '(':
            stack.append(i)
        elif ch == ')':
            if len(stack) > 0:
                stack.pop()
            else:
                invalid.add(i)
        i += 1
    while len(stack) > 0:
        invalid.add(stack.pop())
    j = 0
    while j < l:
        if j not in invalid:
            valid += s[j]
        j += 1
    return valid


s = "(a(b(c)d)"
print(minRemoveToMakeValid(s))
